src/lib/profile/sinks/metricsyaml.cpp

```diff
@@ -122,17 +118,469 @@ static YAML::Emitter& operator<<(YAML::Emitter& e, const MetricScope ms) {
+    }
+    void dump(YAML::Emitter& out) {
+      using namespace YAML;
+      if(!time_x && !time_i) return;
+      out << BeginMap
+          << Key << "name" << Value << "Memory Transfer"
+          << Key << "description" << Value << "Time spent in data transfers to, from and within GPU memory"
+          << Key << "variants" << Value << BeginMap
+            << Key << "Sum" << Value << BeginMap
+              << Key << "render" << Value
+                << Flow << BeginSeq << "number" << "percent" << "colorbar" << EndSeq
+              << Key << "formula" << Value << "sum"
+            << EndMap
+          << EndMap
+          << Key << "children" << BeginSeq;
+      if(time_x) {
+        out << BeginMap
+            << Key << "name" << Value << "Explicit Memory Transfer"
+            << Key << "description" << "Time spent in data transfers explicitly performed by the user"
+            << Key << "variants" << Value << BeginMap;
+        for(const auto& s: time_x->statistics()) {
+          out << Key << s.suffix() << Value << BeginMap
+              << Key << "render" << Value
+                << Flow << BeginSeq << "number" << "hidden";
+          if(s.showPercent()) out << "percent";
+          out << EndSeq
+              << Key << "formula" << Value << BeginMap;
+          rawFormula(out, *time_x, s, MetricScope::execution, "inclusive");
+          if(!rawFormula(out, *time_x, s, MetricScope::function, "exclusive"))
+            rawFormula(out, *time_x, s, MetricScope::point, "point");
+          out << EndMap
+              << EndMap;
+        }
+        out << EndMap
+            << Key << "children" << Value << BeginSeq;
+        for(const Metric& sm: sub_x) rawLeaf(out, sm);
+        out << EndSeq
+            << EndMap;
+      }
+      if(time_i) {
+        out << BeginMap
+            << Key << "name" << Value << "Implicit Memory Transfer"
+            << Key << "description" << "Time spent in data transfers implicitly required by the runtime"
+            << Key << "variants" << Value << BeginMap;
+        for(const auto& s: time_i->statistics()) {
+          out << Key << s.suffix() << Value << BeginMap
+              << Key << "render" << Value
+                << Flow << BeginSeq << "number" << "hidden";
+          if(s.showPercent()) out << "percent";
+          out << EndSeq
+              << Key << "formula" << Value << BeginMap;
+          rawFormula(out, *time_i, s, MetricScope::execution, "inclusive");
+          if(!rawFormula(out, *time_i, s, MetricScope::function, "exclusive"))
+            rawFormula(out, *time_i, s, MetricScope::point, "point");
+          out << EndMap
+              << EndMap;
+        }
+        out << EndMap
+            << Key << "children" << Value << BeginSeq;
+        for(const Metric& sm: sub_i) rawLeaf(out, sm);
+        out << EndSeq
+            << EndMap;
+      }
+      out << EndSeq
+          << EndMap;
+    }
+  } c_memcpy;
+
+  struct MemSet {
```

Comment:
Do those structures need to be hard-coded? Or we can generate code parsing code from a configuration file (i.e., yaml or protocol buffer)?

Reply:
At the moment I haven't worked out how best to dynamically generate the taxonomic structure. I hard-coded the whole thing to get a minimal example together for experimental Viewer support. If you have an idea for an approach please elaborate,  suggestions would be very welcome.